cuqi/implicitprior/_restorator.py

```diff
@@ -0,0 +1,220 @@
+#%%
+from abc import ABC, abstractmethod
+from cuqi.distribution import Distribution
+import numpy as np
+    
+class RestorationPrior(Distribution):
+    """    
+    This class defines an implicit distribution associated with a restoration operator
+    (eg denoiser). They are several works relating restorations operators with
+    priors, see 
+        -Laumont et al. https://arxiv.org/abs/2103.04715
+        -Hu et al. https://openreview.net/pdf?id=x7d1qXEn1e
+    We cannot sample from this distribution, neither compute its logpdf except in
+    some cases. It allows us to apply algorithms such as MYULA and PnPULA.
+    
+    Parameters
+    ---------- 
+    restorator : callable f(x, restoration_strength)
+        Function f that accepts input x to be restored and returns the
+        restored version of x and information about the restoration operation.
+            
+    restorator_kwargs : dictionary
+        Dictionary containing information about the restorator.
+    
+    potential : callable function, optional
+        The potential corresponds to the negative logpdf when it is accessible.
+        This function is a mapping from the parameter domain to the real set. 
+        It can be provided if the user knows how to relate it to the restorator.
+        Ex: restorator is the proximal operator of the total variation (TV), then
+        potential is the TV function. 
+    """
+    def __init__(self, restorator, restorator_kwargs
+                =None, potential=None, **kwargs):
+        if restorator_kwargs is None:
+            restorator_kwargs = {}
+        self.restorator = restorator 
+        self.restorator_kwargs = restorator_kwargs
+        self.potential = potential
+        super().__init__(**kwargs)
+
+    def restore(self, x, restoration_strength):
+        """This function allows us to restore the input x and returns the
+        restored version of x.
+        
+        Parameters
+        ---------- 
+        x : ndarray
+            parameter we want to restore.
+        
+        restoration_strength: positive float
+            Strength of the restoration operation. In the case where the
+            restorator is a denoiser, this parameter might correspond to the
+            noise level.
+        """
+        solution, info = self.restorator(x, restoration_strength=restoration_strength,
+                                         **self.restorator_kwargs)
+        self.info = info
+        return solution
+    
+    def logpdf(self, x):
+        """The logpdf function. It returns nan because we don't know the 
+        logpdf of the implicit prior."""
+        if self.potential == None:
+            return np.nan
+        else:
+            return -self.potential(x)
+    
+    def _sample(self, N, rng=None):
+        raise NotImplementedError("The sample method is not implemented for the"
+                                  + "RestorationPrior class.")
+
+    #TODO this method copied from userdefinedistribution
```

Comment:
Maybe we capture this todo (and the other todos) in an issue instead?

Reply:
Thanks for the comment. We created issue #563